SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp

```diff
@@ -197,3 +169,116 @@ Worker_ComponentId USpatialDispatcher::GetComponentId(Worker_Op* Op) const
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMPONENT_UPDATE, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->component_update);
+	});
+}
+
+USpatialDispatcher::FCallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const TFunction<void(const Worker_CommandRequestOp&)>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMMAND_REQUEST, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->command_request);
+	});
+}
+
+USpatialDispatcher::FCallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const TFunction<void(const Worker_CommandResponseOp&)>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMMAND_RESPONSE, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->command_response);
+	});
+}
+
+USpatialDispatcher::FCallbackId USpatialDispatcher::AddGenericOpCallback(Worker_ComponentId ComponentId, Worker_OpType OpType, const TFunction<void(const Worker_Op*)>& Callback)
+{
+	check(SpatialConstants::MIN_EXTERNAL_SCHEMA_ID <= ComponentId && ComponentId <= SpatialConstants::MAX_EXTERNAL_SCHEMA_ID);
+	const FCallbackId NewCallbackId = NextCallbackId++;
+	ComponentOpTypeToCallbackIdMap.FindOrAdd(ComponentId).FindOrAdd(OpType).Add(UserOpCallbackData{ NewCallbackId, Callback });
+	CallbackIdToDataMap.Add(NewCallbackId, CallbackIdData{ ComponentId, OpType });
+	return NewCallbackId;
+}
+
+void USpatialDispatcher::RemoveOpCallback(FCallbackId CallbackId)
+{
+	// Find callback ID in map and assert if it does not exist
+	CallbackIdData CallbackDataToRemove = CallbackIdToDataMap.FindAndRemoveChecked(CallbackId);
+
+	TMap<Worker_OpType, TArray<UserOpCallbackData>>* OpTypesToCallbackIds = ComponentOpTypeToCallbackIdMap.Find(CallbackDataToRemove.ComponentId);
+	check(OpTypesToCallbackIds);
+	TArray<UserOpCallbackData>* ComponentCallbackIds = OpTypesToCallbackIds->Find(CallbackDataToRemove.OpType);
+	check(ComponentCallbackIds);
+	UserOpCallbackData* CallbackData = ComponentCallbackIds->FindByPredicate([CallbackId](const UserOpCallbackData& Data)
+	{
+		return Data.Id == CallbackId;
+	});
+	check(CallbackData);
+
+	// If removing only callback for a component ID / op type, delete map entries as applicable
+	if (OpTypesToCallbackIds->Num() == 1)
+	{
+		if (ComponentOpTypeToCallbackIdMap.Num() == 1)
+		{
+			ComponentOpTypeToCallbackIdMap.Remove(CallbackDataToRemove.ComponentId);
+			return;
+		}
+		OpTypesToCallbackIds->Remove(CallbackDataToRemove.OpType);
+		return;
+	}
+
+	for (int i = 0; i < ComponentCallbackIds->Num(); i++) {
+		if ((*ComponentCallbackIds)[i].Id == CallbackId)
+		{
+			ComponentCallbackIds->RemoveAt(i);
+			return;
+		}
+	}
+
+}
+
+void USpatialDispatcher::RunCallbacks(Worker_ComponentId ComponentId, const Worker_Op* Op)
+{
+	TMap<Worker_OpType, TArray<UserOpCallbackData>>* OpTypeCallbacks = ComponentOpTypeToCallbackIdMap.Find(ComponentId);
```

Comment:
Similar to Samis comment below, `ComponentOpTypeToCallbackIdMap` should probably be ComponentOpTypeToCallbackMap